nvflare/app_common/abstract/fl_model.py

```diff
@@ -0,0 +1,78 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from enum import Enum
+from typing import Dict, Optional
+
+
+class TransferType(Enum):
+    MODEL = "MODEL"
+    MODEL_DIFF = "MODEL_DIFF"
+
+
+class FLModelConst:
+    AGGREGATION = "aggregation"
+    METRICS = "metrics"
+
+
+class FLModel:
+    def __init__(
+        self,
+        transfer_type: TransferType,
+        model: Dict,
+        optimizer: Optional[Dict] = None,
+        metrics: Optional[Dict] = None,
+        configs: Optional[Dict] = None,
+        client_weights: Optional[Dict] = None,
+        round: Optional[int] = None,
+        meta: Optional[Dict] = None,
+    ):
+        """
+        Args:
+            transfer_type: how the model will be transferred: as whole model (such as weight) or model_diff (weight_diff)
+            model:  machine learning model, for Deep learning, this could be weights such pytorch.state_dict or others depending on type of model
+            optimizer: optionally provided optimizer, for many cases, this optimizer doesn't need to be transfer between FL training.
+            metrics: evaluation metrics such as loss and scores
+            configs: training configurations that is dynamically changes during training and need to be passed around.
```

Comment:
Mostly, config is a one-time information that needs to be shared/loaded. For dynamical changes, `meta` could be utilized.

Reply:
Both are rare use cases. Meta could be more flexible as it could be hyperparameters, data, or both.

Reply:
Does "config" really have more structure?

To ensure / enforce structure, we could add some validation to the keys of config.
But same can be applied to meta?

Another thing is that, what "key" do we expect to exist in "configs"?

Reply:
So, "initial_learning_rate" should be put inside config while "current_learning_rate" (as it might be dynamic) should be put inside meta?

Reply:
So if we have "config", it can contain both fixed stuff (initial learning rate, decay schedule) and dynamic stuff (current learning rate)

And the dynamic stuff does not always exist as some examples don't need to pass current learning rate around.

Reply:
I guess it is the name that is confusing to me.

If we say "static" config should stay OUT of FLModel, and it is a one-time broadcast thing.

Then this name inside the FLModel maybe we name it "dynamic_configs" or some better names so people knows that the regular/static config does NOT need to be in here.